<div style="background:#f8fafc; border:1px solid #dbeafe; border-left:7px solid #2563eb; padding:26px 30px; border-radius:16px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<h1 style="margin-top:0; color:#1d4ed8; font-size:34px;">
📊 01 Data Exploration
</h1>

<p style="font-size:17px; line-height:1.8; color:#334155;">
This notebook explores the raw CSV files from the 
<b>Open University Learning Analytics Dataset (OULAD)</b> and provides an initial understanding of the data structure before building the machine-learning pipeline.
</p>

<p style="font-size:17px; line-height:1.8; color:#334155;">
The OULAD context is important because the records describe <b>distance-learning students</b>. VLE clicks, assessment submissions, registration dates, demographics, and final outcomes can be analyzed together to understand student engagement, withdrawal risk, and academic performance in online learning.
</p>

<hr style="border:0; border-top:1px solid #bfdbfe; margin:24px 0;">

<h2 style="color:#0f172a; font-size:26px;">
🎯 Main Dataset Granularity
</h2>

<p style="font-size:17px; line-height:1.7; color:#334155;">
The final machine-learning dataset is designed at the following grain:
</p>

<div style="background:#ffffff; border:1px solid #bfdbfe; border-radius:14px; padding:18px 22px; margin:14px 0 22px 0; box-shadow:0 2px 8px rgba(37,99,235,0.08);">

<p style="font-size:16px; margin-top:0; color:#1e40af;"><b>One row per:</b></p>

<ul style="font-size:16px; line-height:2; margin-bottom:0;">
  <li><code style="background:#e0f2fe; color:#075985; padding:3px 7px; border-radius:6px;">id_student</code></li>
  <li><code style="background:#e0f2fe; color:#075985; padding:3px 7px; border-radius:6px;">code_module</code></li>
  <li><code style="background:#e0f2fe; color:#075985; padding:3px 7px; border-radius:6px;">code_presentation</code></li>
</ul>

</div>

<p style="font-size:17px; line-height:1.8; color:#334155;">
This structure prevents duplicated student records caused by repeated interactions and creates a clean 
<b>student-level dataset</b> for predictive modeling.
</p>

</div>

<div style="background:#f8fafc; border:1px solid #dbeafe; border-left:7px solid #0f766e; padding:26px 30px; border-radius:16px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<h1 style="margin-top:0; color:#0f766e; font-size:32px;">
📘 Data Dictionary
</h1>

<p style="font-size:17px; line-height:1.8; color:#334155;">
This section summarizes the main columns used throughout the project and explains their meaning in the 
<b>OULAD dataset</b>.
</p>

<hr style="border:0; border-top:1px solid #99f6e4; margin:24px 0;">

<h2 style="color:#0f766e;">📚 Course and Presentation Columns</h2>

<ul style="font-size:16px; line-height:2;">
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">code_module</code>: module identifier.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">code_presentation</code>: presentation identifier, such as <code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">2013B</code> for February or <code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">2013J</code> for October.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">module_presentation_length</code>: length of the module presentation in days.</li>
</ul>

<h2 style="color:#0f766e;">👤 Student Information Columns</h2>

<ul style="font-size:16px; line-height:2;">
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">id_student</code>: student identifier.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">gender</code>: student gender.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">region</code>: student geographic region.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">highest_education</code>: highest education level achieved by the student before the course.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">imd_band</code>: socio-economic deprivation band of the student’s area.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">age_band</code>: student age group.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">disability</code>: whether the student declared a disability.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">num_of_prev_attempts</code>: previous attempts by the student for the module.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">studied_credits</code>: total credits the student was studying at the time.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">final_result</code>: final student outcome for the module presentation.</li>
</ul>

<h2 style="color:#0f766e;">📝 Registration Columns</h2>

<ul style="font-size:16px; line-height:2;">
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">date_registration</code>: registration date relative to the module start.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">date_unregistration</code>: unregistration date relative to the module start.</li>
</ul>

<h2 style="color:#0f766e;">🧪 Assessment Columns</h2>

<ul style="font-size:16px; line-height:2;">
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">id_assessment</code>: assessment identifier.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">assessment_type</code>: assessment type: Tutor Marked Assessment (<code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">TMA</code>), Computer Marked Assessment (<code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">CMA</code>), or <code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">Exam</code>.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">date</code>: relative day number from the start of the module presentation. In <code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">assessments</code>, this is the assessment due date.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">weight</code>: assessment weight as a percentage.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">date_submitted</code>: student submission date, relative to the start of the module presentation.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">is_banked</code>: whether an assessment result was transferred from a previous presentation.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">score</code>: assessment score from 0 to 100; scores below 40 are considered failing scores.</li>
</ul>

<h2 style="color:#0f766e;">🖱️ VLE Interaction Columns</h2>

<ul style="font-size:16px; line-height:2;">
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">id_site</code>: VLE material identifier.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">date</code>: relative day number from the start of the module presentation. In <code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">studentVle</code>, this is the interaction date.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">sum_click</code>: number of interactions with a VLE material on a day.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">activity_type</code>: VLE material role or category.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">week_from</code>: planned start week for using the VLE material.</li>
  <li><code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">week_to</code>: planned end week for using the VLE material.</li>
</ul>

<div style="background:#ecfeff; border:1px solid #67e8f9; border-radius:12px; padding:16px 18px; margin-top:22px;">

<p style="margin:0; font-size:16px; line-height:1.7; color:#164e63;">
<b>Note:</b><br>
The column <code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">date</code> appears in more than one table. In <code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">studentVle</code>, it represents the interaction date, while in <code style="background:#e0f2fe; color:#0369a1; padding:3px 8px; border-radius:6px;">assessments</code>, it represents the assessment due date.
</p>

</div>

</div>

<h2 style="color:#EA580C;">📁 Load Dataset Paths</h2>

In [1]:
from pathlib import Path
import pandas as pd

# Resolve project paths from this notebook location when possible.
# The notebook lives in project/backend/notebooks, so project root is two levels up.
try:
    NOTEBOOK_DIR = Path(__file__).resolve().parent
except NameError:
    NOTEBOOK_DIR = Path.cwd().resolve()

path_candidates = [
    NOTEBOOK_DIR.parents[1] if len(NOTEBOOK_DIR.parents) > 1 else NOTEBOOK_DIR,  # when cwd is project/backend/notebooks
    NOTEBOOK_DIR,                     # when cwd is project
    NOTEBOOK_DIR / "project",       # when cwd is repository root
    Path.cwd().resolve(),
    Path.cwd().resolve() / "project",
]

PROJECT_DIR = None
for candidate in path_candidates:
    if (candidate / "data" / "raw").exists():
        PROJECT_DIR = candidate
        break

if PROJECT_DIR is None:
    raise FileNotFoundError(
        "Could not locate project/data/raw. Run this notebook from the project folder "
        "or from project/backend/notebooks."
    )

# Define the path to the raw OULAD CSV files.
RAW_DIR = PROJECT_DIR / "data" / "raw"

# Define the path where processed datasets will be saved.
PROCESSED_DIR = PROJECT_DIR / "data" / "processed"
PROCESSED_DIR.mkdir(parents=True, exist_ok=True)

# Collect all CSV files available in the raw data directory.
csv_files = sorted(RAW_DIR.glob("*.csv"))

print("Project directory:", PROJECT_DIR)
print("Raw data directory:", RAW_DIR)
print("CSV files found:", len(csv_files))

# Display the discovered CSV files to verify that the dataset files are accessible.
csv_files


Project directory: C:\Users\HP\OneDrive\المستندات\GitHub\Student-Leak-Radar\project
Raw data directory: C:\Users\HP\OneDrive\المستندات\GitHub\Student-Leak-Radar\project\data\raw
CSV files found: 6


[WindowsPath('C:/Users/HP/OneDrive/المستندات/GitHub/Student-Leak-Radar/project/data/raw/assessments.csv'),
 WindowsPath('C:/Users/HP/OneDrive/المستندات/GitHub/Student-Leak-Radar/project/data/raw/courses.csv'),
 WindowsPath('C:/Users/HP/OneDrive/المستندات/GitHub/Student-Leak-Radar/project/data/raw/studentAssessment.csv'),
 WindowsPath('C:/Users/HP/OneDrive/المستندات/GitHub/Student-Leak-Radar/project/data/raw/studentInfo.csv'),
 WindowsPath('C:/Users/HP/OneDrive/المستندات/GitHub/Student-Leak-Radar/project/data/raw/studentRegistration.csv'),
 WindowsPath('C:/Users/HP/OneDrive/المستندات/GitHub/Student-Leak-Radar/project/data/raw/studentVle.csv')]

<h2 style="color:#EA580C;">🔎 Inspect Files, Columns, and Row Counts</h2>

In [2]:
overview = []

if not csv_files:
    raise FileNotFoundError(
        f"No CSV files found in {RAW_DIR}. Check that the OULAD raw CSV files are in project/data/raw."
    )

# Loop through each raw CSV file and collect basic metadata.
for path in csv_files:
    sample = pd.read_csv(path, nrows=5)

    # Count data rows without loading the full file into memory.
    with open(path, "rb") as f:
        rows = sum(1 for _ in f) - 1

    overview.append({
        "file": path.name,
        "rows": rows,
        "columns_count": len(sample.columns),
        "columns": ", ".join(sample.columns)
    })

# Display a compact overview of all raw dataset files.
pd.DataFrame(overview).sort_values("file")


,file,rows,columns_count,columns
0,assessments.csv,206,6,"code_module, code_presentation, id_assessment,..."
1,courses.csv,22,3,"code_module, code_presentation, module_present..."
2,studentAssessment.csv,173912,5,"id_assessment, id_student, date_submitted, is_..."
3,studentInfo.csv,32593,12,"code_module, code_presentation, id_student, ge..."
4,studentRegistration.csv,32593,5,"code_module, code_presentation, id_student, da..."
5,studentVle.csv,10655280,6,"code_module, code_presentation, id_student, id..."


<h2 style="color:#EA580C;">👀 Preview Raw Dataset Samples</h2>

In [3]:
overview[1]

{'file': 'courses.csv',
 'rows': 22,
 'columns_count': 3,
 'columns': 'code_module, code_presentation, module_presentation_length'}

In [4]:
# Display the first five rows of each raw CSV file. This helps us quickly understand the structure and sample values of every dataset.
for path in csv_files:
    print(f"\n{path.name}")
    display(pd.read_csv(path, nrows=5))


assessments.csv


,code_module,code_presentation,id_assessment,assessment_type,date,weight
0,AAA,2013J,1752,TMA,19,10
1,AAA,2013J,1753,TMA,54,20
2,AAA,2013J,1754,TMA,117,20
3,AAA,2013J,1755,TMA,166,20
4,AAA,2013J,1756,TMA,215,30



courses.csv


,code_module,code_presentation,module_presentation_length
0,AAA,2013J,268
1,AAA,2014J,269
2,BBB,2013J,268
3,BBB,2014J,262
4,BBB,2013B,240



studentAssessment.csv


,id_assessment,id_student,date_submitted,is_banked,score
0,1752,11391,18,0,78
1,1752,28400,22,0,70
2,1752,31604,17,0,72
3,1752,32885,26,0,69
4,1752,38053,19,0,79



studentInfo.csv


,code_module,code_presentation,id_student,gender,region,highest_education,imd_band,age_band,num_of_prev_attempts,studied_credits,disability,final_result
0,AAA,2013J,11391,M,East Anglian Region,HE Qualification,90-100%,55<=,0,240,N,Pass
1,AAA,2013J,28400,F,Scotland,HE Qualification,20-30%,35-55,0,60,N,Pass
2,AAA,2013J,30268,F,North Western Region,A Level or Equivalent,30-40%,35-55,0,60,Y,Withdrawn
3,AAA,2013J,31604,F,South East Region,A Level or Equivalent,50-60%,35-55,0,60,N,Pass
4,AAA,2013J,32885,F,West Midlands Region,Lower Than A Level,50-60%,0-35,0,60,N,Pass



studentRegistration.csv


,code_module,code_presentation,id_student,date_registration,date_unregistration
0,AAA,2013J,11391,-159,NaN
1,AAA,2013J,28400,-53,NaN
2,AAA,2013J,30268,-92,12.0
3,AAA,2013J,31604,-52,NaN
4,AAA,2013J,32885,-176,NaN



studentVle.csv


,code_module,code_presentation,id_student,id_site,date,sum_click
0,AAA,2013J,28400,546652,-10,4
1,AAA,2013J,28400,546652,-10,1
2,AAA,2013J,28400,546652,-10,1
3,AAA,2013J,28400,546614,-10,11
4,AAA,2013J,28400,546714,-10,1


<h2 style="color:#EA580C;">📥 Load Core Dataset Tables</h2>

`studentVle.csv` is large, so it will be loaded later only when needed for VLE aggregation.

In [5]:
# Load the core OULAD tables needed for exploration and later merging.
courses = pd.read_csv(RAW_DIR / "courses.csv")
assessments = pd.read_csv(RAW_DIR / "assessments.csv")
student_info = pd.read_csv(RAW_DIR / "studentInfo.csv")
student_registration = pd.read_csv(RAW_DIR / "studentRegistration.csv")
student_assessment = pd.read_csv(RAW_DIR / "studentAssessment.csv")

# vle.csv is useful for activity-type labels, but the rest of the exploration can continue without it.
vle_path = RAW_DIR / "vle.csv"
if vle_path.exists():
    vle = pd.read_csv(vle_path)
else:
    print(f"Warning: {vle_path} was not found. VLE activity-type labels will be unavailable.")
    vle = pd.DataFrame(columns=[
        "id_site", "code_module", "code_presentation", "activity_type", "week_from", "week_to"
    ])

# Store loaded tables in a dictionary for quick inspection.
tables = {
    "courses": courses,
    "assessments": assessments,
    "student_info": student_info,
    "student_registration": student_registration,
    "student_assessment": student_assessment,
    "vle": vle,
}

# Display the shape of each loaded table.
pd.DataFrame(
    {name: df.shape for name, df in tables.items()},
    index=["rows", "columns"]
).T


,rows,columns
courses,22,3
assessments,206,6
student_info,32593,12
student_registration,32593,5
student_assessment,173912,5
vle,0,6


<h2 style="color:#EA580C;">📊 Basic Exploratory Charts</h2>

This section visualizes the main structure of the dataset, including student distribution, module distribution, attempts, and final results.

In [6]:
# Calculate basic dataset statistics.
total_rows = len(student_info)
unique_students = student_info["id_student"].nunique()
modules = sorted(student_info["code_module"].unique())

# Build a compact summary table for the main dataset structure.
summary = pd.DataFrame({
    "metric": [
        "student_module_presentation_rows",
        "unique_students",
        "number_of_modules"
    ],
    "value": [
        total_rows,
        unique_students,
        len(modules)
    ]
})

# Display the dataset summary metrics.
summary

,metric,value
0,student_module_presentation_rows,32593
1,unique_students,28785
2,number_of_modules,7


<h2 style="color:#EA580C;">👨‍🎓 Student Distribution Across Modules</h2>

This section calculates the number of unique students enrolled in each module.

A student may appear multiple times in the dataset because the same module can have multiple presentations. Therefore, we count unique student IDs instead of total rows.

In [7]:
# Count unique students enrolled in each module.
students_by_module = (
    student_info
    .groupby("code_module")["id_student"]
    .nunique()
    .sort_values(ascending=False)
    .reset_index(name="unique_students")
)

# Display the number of unique students per module.
students_by_module

,code_module,unique_students
0,BBB,7692
1,FFF,7397
2,DDD,5848
3,CCC,4251
4,EEE,2859
5,GGG,2525
6,AAA,712


<h2 style="color:#EA580C;">🧾 Student Attempts Across Modules</h2>

This section counts student-module-presentation rows for each module.

Unlike the previous section, this represents total enrollment attempts, not unique students. A student can appear more than once if they enrolled in multiple presentations of the same module.

In [8]:
# Count the number of student-module-presentation rows for each module.
# This represents enrollment attempts rather than unique students.
attempts_by_module = (
    student_info["code_module"]
    .value_counts()
    .rename_axis("code_module")
    .reset_index(name="student_module_presentation_rows")
)

# Display total attempts per module.
attempts_by_module

,code_module,student_module_presentation_rows
0,BBB,7909
1,FFF,7762
2,DDD,6272
3,CCC,4434
4,EEE,2934
5,GGG,2534
6,AAA,748


In [9]:
import plotly.express as px

# Create a bar chart to compare total enrollment attempts across modules.
fig = px.bar(
    attempts_by_module,
    x="code_module",
    y="student_module_presentation_rows",
    text="student_module_presentation_rows",
    title="Student-Module-Presentation Rows by Module"
)

# Customize bar color and show values above each bar.
fig.update_traces(
    marker_color="#10b981",
    textposition="outside"
)

# Improve chart layout for readability.
fig.update_layout(
    xaxis_title="Module",
    yaxis_title="Rows",
    width=800,
    height=500,
    showlegend=False
)

fig.show()

<div style="background:#fff7ed; border:1px solid #fdba74; border-left:5px solid #ea580c; padding:16px 20px; border-radius:12px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<p style="font-size:16px; margin-top:0; color:#c2410c;">
<b>🔎 Observation</b>
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
Modules 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">BBB</code> 
and 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">FFF</code> 
contain the highest number of student-module-presentation records, while module 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">AAA</code> 
has the smallest number of records.
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155; margin-bottom:0;">
This shows that the dataset is not evenly distributed across modules. This imbalance should be considered later when comparing model performance across modules or interpreting prediction results.
</p>

</div>

<h2 style="color:#EA580C;">📚 Students Enrolled in Multiple Modules</h2>

This section analyzes how many unique modules each student enrolled in across the dataset.

In [10]:
# Count how many unique modules each student enrolled in.
modules_per_student = (
    student_info
    .groupby("id_student")["code_module"]
    .nunique()
    .reset_index(name="number_of_modules")
)

# Count students who enrolled in more than one module.
students_more_than_one_module = (
    modules_per_student["number_of_modules"] > 1
).sum()

students_more_than_one_module

np.int64(2479)

In [11]:
# Calculate the distribution of students by number of enrolled modules.
students_by_module_count = (
    modules_per_student["number_of_modules"]
    .value_counts()
    .sort_index()
    .rename_axis("number_of_modules")
    .reset_index(name="students")
)

# Display how many students enrolled in 1, 2, 3, or more modules.
students_by_module_count

,number_of_modules,students
0,1,26306
1,2,2459
2,3,20


<div style="background:#fff7ed; border:1px solid #fdba74; border-left:5px solid #ea580c; padding:16px 20px; border-radius:12px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<p style="font-size:16px; margin-top:0; color:#c2410c;">
<b>🔎 Observation</b>
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155; margin-bottom:0;">
Most students are enrolled in only one module, while a smaller group appears in multiple modules. This is important because student behavior and outcomes may vary across modules and presentations, which can affect feature engineering and later predictive analysis.
</p>

</div>

In [12]:
import plotly.express as px

# Create a bar chart showing how many students took 1, 2, or more modules.
fig = px.bar(
    students_by_module_count,
    x="number_of_modules",
    y="students",
    text="students",
    title="Students by Number of Modules Taken"
)

# Customize bar color and display values above the bars.
fig.update_traces(
    marker_color="#f59e0b",
    textposition="outside"
)

# Improve chart readability.
fig.update_layout(
    xaxis_title="Number of modules",
    yaxis_title="Students",
    width=700,
    height=500,
    showlegend=False
)

fig.show()

<div style="background:#fff7ed; border:1px solid #fdba74; border-left:5px solid #ea580c; padding:16px 20px; border-radius:12px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<p style="font-size:16px; margin-top:0; color:#c2410c;">
<b>🔎 Observation</b>
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
Most students are enrolled in only one module, while a much smaller number of students are enrolled in two modules. Very few students appear in three modules.
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155; margin-bottom:0;">
This shows that the dataset is mainly dominated by single-module student records, which supports using 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">id_student</code>, 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">code_module</code>, 
and 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">code_presentation</code> 
together as the main unit of analysis.
</p>

</div>

<h2 style="color:#EA580C;">📈 Module Outcomes by Final Result</h2>

This section compares final student outcomes across different modules.

In [13]:
# Build a cross-tabulation between modules and final student outcomes.
result_by_module = pd.crosstab(
    student_info["code_module"],
    student_info["final_result"]
)

# Display the distribution of final results for each module.
result_by_module

final_result,Distinction,Fail,Pass,Withdrawn
code_module,,,,
AAA,44,91,487,126
BBB,677,1767,3077,2388
CCC,498,781,1180,1975
DDD,383,1412,2227,2250
EEE,356,562,1294,722
FFF,670,1711,2978,2403
GGG,396,728,1118,292


In [14]:
import plotly.express as px

# Create a stacked bar chart to compare final outcomes across modules.
fig = px.bar(
    result_by_module.reset_index(),
    x="code_module",
    y=result_by_module.columns,
    title="Final Result by Module"
)

# Use stacked bars to show the composition of outcomes within each module.
fig.update_layout(
    xaxis_title="Module",
    yaxis_title="Student-module-presentation rows",
    width=900,
    height=500,
    barmode="stack"
)

fig.show()

<div style="background:#fff7ed; border:1px solid #fdba74; border-left:5px solid #ea580c; padding:16px 20px; border-radius:12px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<p style="font-size:16px; margin-top:0; color:#c2410c;">
<b>🔎 Observation</b>
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
Modules 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">BBB</code> 
and 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">FFF</code> 
have the largest overall number of student-module-presentation records, and they also show high numbers of both successful and at-risk outcomes.
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155; margin-bottom:0;">
The stacked distribution shows that final outcomes are not identical across modules, which means module-level differences should be considered when interpreting model results.
</p>

</div>

<h2 style="color:#EA580C;">🧩 Missing Values Summary</h2>

This section checks missing values across the loaded dataset tables to identify columns that may need cleaning, imputation, or special handling before modeling.

In [15]:
missing_summary = []

# Loop through each loaded table and calculate missing values per column.
for name, df in tables.items():
    missing = df.isna().sum()
    missing = missing[missing > 0]

    # Store only columns that contain missing values.
    for column, count in missing.items():
        missing_summary.append({
            "table": name,
            "column": column,
            "missing_count": int(count),
            "missing_pct": round(count / len(df) * 100, 2)
        })

# Display missing-value summary sorted by table and column name.
pd.DataFrame(missing_summary).sort_values(["table", "column"])

,table,column,missing_count,missing_pct
0,assessments,date,11,5.34
4,student_assessment,score,173,0.10
1,student_info,imd_band,1111,3.41
2,student_registration,date_registration,45,0.14
3,student_registration,date_unregistration,22521,69.10


<div style="background:#fff7ed; border:1px solid #fdba74; border-left:5px solid #ea580c; padding:16px 20px; border-radius:12px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<p style="font-size:16px; margin-top:0; color:#c2410c;">
<b>🔎 Observation</b>
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
Most missing values are concentrated in a small number of columns. The largest missing percentage appears in 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">date_unregistration</code>, 
which is expected because many students did not unregister from the course.
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155; margin-bottom:0;">
Other missing values, such as 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">score</code>, 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">date_registration</code>, 
and 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">imd_band</code>, 
are relatively limited and can be handled later during preprocessing.
</p>

</div>

<h2 style="color:#EA580C;">🏷️ Categorical Value Distributions</h2>

This section reviews the value distribution of important categorical columns, such as demographics, assessment types, VLE activity types, and final outcomes.

In [16]:
# Define the categorical columns to inspect from each table.
categorical_columns = {
    "student_info": [
        "gender",
        "region",
        "highest_education",
        "imd_band",
        "age_band",
        "disability",
        "final_result"
    ],
    "assessments": ["assessment_type"],
    "vle": ["activity_type"],
}

categorical_summary = []

for table_name, columns in categorical_columns.items():
    df = tables[table_name]

    for column in columns:
        counts = df[column].value_counts(dropna=False)
        percentages = df[column].value_counts(dropna=False, normalize=True) * 100

        for category, count in counts.items():
            categorical_summary.append({
                "table": table_name,
                "column": column,
                "category": category,
                "count": count,
                "percentage": round(percentages.loc[category], 2)
            })

categorical_summary_df = pd.DataFrame(categorical_summary)

categorical_summary_df

,table,column,category,count,percentage
0,student_info,gender,M,17875,54.84
1,student_info,gender,F,14718,45.16
2,student_info,region,Scotland,3446,10.57
3,student_info,region,East Anglian Region,3340,10.25
4,student_info,region,London Region,3216,9.87
5,student_info,region,South Region,3092,9.49
6,student_info,region,North Western Region,2906,8.92
7,student_info,region,West Midlands Region,2582,7.92
8,student_info,region,South West Region,2436,7.47
9,student_info,region,East Midlands Region,2365,7.26


<h2 style="color:#EA580C;">🔗 Relationship and Key Integrity Checks</h2>

This section verifies the main relational keys used across the OULAD tables and checks whether important identifiers remain unique where expected.

### Important Keys

- **Course presentation key:**  
  `code_module`, `code_presentation`

- **Student attempt key:**  
  `id_student`, `code_module`, `code_presentation`

- **Assessment key:**  
  `id_assessment`

- **VLE site key:**  
  `id_site`, `code_module`, `code_presentation`

In [17]:
# Define the main relational keys used across the dataset.
student_key = ["id_student", "code_module", "code_presentation"]
course_key = ["code_module", "code_presentation"]
site_key = ["id_site", "code_module", "code_presentation"]

# Check whether important keys remain unique where expected.
checks = {
    "student_info rows": len(student_info),
    "student_info unique student attempts":
        student_info[student_key].drop_duplicates().shape[0],

    "student_registration unique student attempts":
        student_registration[student_key].drop_duplicates().shape[0],

    "courses rows": len(courses),
    "courses unique course presentations":
        courses[course_key].drop_duplicates().shape[0],

    "assessments unique assessment ids":
        assessments["id_assessment"].nunique(),

    "vle unique site keys":
        vle[site_key].drop_duplicates().shape[0],
}

# Display relationship and uniqueness checks.
pd.Series(checks).to_frame("value")

,value
student_info rows,32593
student_info unique student attempts,32593
student_registration unique student attempts,32593
courses rows,22
courses unique course presentations,22
assessments unique assessment ids,206
vle unique site keys,0


<div style="background:#fff7ed; border:1px solid #fdba74; border-left:5px solid #ea580c; padding:16px 20px; border-radius:12px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<p style="font-size:16px; margin-top:0; color:#c2410c;">
<b>🔎 Observation</b>
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
The key integrity checks confirm that the expected identifiers remain unique across the main tables.
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155; margin-bottom:0;">
In particular, the combination of 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">id_student</code>, 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">code_module</code>, 
and 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">code_presentation</code> 
uniquely represents student attempts, which supports using this structure as the main analysis grain for the project.
</p>

</div>

<h2 style="color:#EA580C;">📝 Build Assessment Features</h2>

This section merges student assessment submissions with assessment metadata, then aggregates assessment behavior at the student-course-presentation level.

In [18]:
# Merge student submissions with assessment metadata.
# This adds module, presentation, due date, assessment type, and weight information.
assessment_events = student_assessment.merge(
    assessments,
    on="id_assessment",
    how="left",
    validate="many_to_one"
)

# Create assessment-level indicators before aggregation.
assessment_events["weighted_score_component"] = (
    assessment_events["score"] * assessment_events["weight"]
)
assessment_events["late_submission"] = (
    assessment_events["date_submitted"] > assessment_events["date"]
)

# Aggregate assessment behavior at the student-course-presentation level.
assessment_features = (
    assessment_events
    .groupby(student_key, as_index=False)
    .agg(
        assessment_submissions=("id_assessment", "count"),
        assessment_score_mean=("score", "mean"),
        assessment_score_min=("score", "min"),
        assessment_score_max=("score", "max"),
        assessment_weight_sum=("weight", "sum"),
        assessment_weighted_score_sum=("weighted_score_component", "sum"),
        late_submissions=("late_submission", "sum"),
        banked_assessments=("is_banked", "sum"),
    )
)

# Calculate the weighted average assessment score.
assessment_features["assessment_weighted_score_mean"] = (
    assessment_features["assessment_weighted_score_sum"] /
    assessment_features["assessment_weight_sum"]
)

# Preview the generated assessment-level features.
assessment_features.head()

,id_student,code_module,code_presentation,assessment_submissions,assessment_score_mean,assessment_score_min,assessment_score_max,assessment_weight_sum,assessment_weighted_score_sum,late_submissions,banked_assessments,assessment_weighted_score_mean
0,6516,AAA,2014J,5,61.800000,48.0,77.0,100.0,6350.0,0,0,63.50
1,8462,DDD,2013J,3,87.666667,83.0,93.0,40.0,3490.0,1,0,87.25
2,8462,DDD,2014J,4,86.500000,83.0,93.0,50.0,4300.0,0,4,86.00
3,11391,AAA,2013J,5,82.000000,78.0,85.0,100.0,8240.0,0,0,82.40
4,23629,BBB,2013B,4,82.500000,63.0,100.0,25.0,1669.0,3,0,66.76


<div style="background:#fff7ed; border:1px solid #fdba74; border-left:5px solid #ea580c; padding:16px 20px; border-radius:12px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<p style="font-size:16px; margin-top:0; color:#c2410c;">
<b>🔎 Observation</b>
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
The aggregated assessment features provide a compact student-level summary of assessment behavior for each module presentation.
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
These features capture important academic signals such as assessment participation, average and weighted scores, late submissions, assessment difficulty exposure through weights, and transferred 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">banked</code> assessments.
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
The final table now follows the main project granularity:
</p>

<ul style="font-size:15.5px; line-height:2; color:#334155;">
  <li>One row per:
    <ul style="line-height:2; margin-top:6px;">
      <li><code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">id_student</code></li>
      <li><code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">code_module</code></li>
      <li><code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">code_presentation</code></li>
    </ul>
  </li>
</ul>

<p style="font-size:15.5px; line-height:1.8; color:#334155; margin-bottom:0;">
This structure is suitable for later machine learning and early-warning prediction tasks.
</p>

</div>

<h2 style="color:#EA580C;">🧪 Build Assessment-Type Features</h2>

This section creates separate features for each assessment type, such as `TMA`, `CMA`, and `Exam`, to capture how student performance differs across assessment formats.

In [19]:
# Create assessment-type features using pivot table.
# This separates count and mean score by assessment type.
assessment_type_features = (
    assessment_events
    .pivot_table(
        index=student_key,
        columns="assessment_type",
        values="score",
        aggfunc=["count", "mean"],
        fill_value=0
    )
)

# Flatten multi-level column names into readable feature names.
assessment_type_features.columns = [
    f"assessment_{metric.lower()}_{assessment_type.lower()}"
    for metric, assessment_type in assessment_type_features.columns
]

# Reset index to return student identifiers as regular columns.
assessment_type_features = assessment_type_features.reset_index()

# Preview assessment-type-level features.
assessment_type_features.head()

,id_student,code_module,code_presentation,assessment_count_cma,assessment_count_exam,assessment_count_tma,assessment_mean_cma,assessment_mean_exam,assessment_mean_tma
0,6516,AAA,2014J,0,0,5,0.0,0.0,61.800000
1,8462,DDD,2013J,0,0,3,0.0,0.0,87.666667
2,8462,DDD,2014J,0,0,4,0.0,0.0,86.500000
3,11391,AAA,2013J,0,0,5,0.0,0.0,82.000000
4,23629,BBB,2013B,2,0,2,100.0,0.0,65.000000


<div style="background:#fff7ed; border:1px solid #fdba74; border-left:5px solid #ea580c; padding:16px 20px; border-radius:12px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<p style="font-size:16px; margin-top:0; color:#c2410c;">
<b>🔎 Observation</b>
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
The assessment-type features separate student performance by assessment format, such as 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">TMA</code>, 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">CMA</code>, 
and 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">Exam</code>.
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155; margin-bottom:0;">
This adds more detailed academic signals than using overall assessment averages only, because different assessment types may reflect different learning behaviors and performance patterns.
</p>

</div>

<h2 style="color:#EA580C;">🖱️ Build VLE Click Features</h2>

This section builds student engagement features from VLE clickstream data.  
Because `studentVle.csv` is the largest table, it is loaded only when needed, then joined with `vle.csv` to add activity-type information before aggregation.

In [20]:
# Load the large student VLE interaction table only when needed.
student_vle = pd.read_csv(RAW_DIR / "studentVle.csv")

# Merge VLE interactions with VLE metadata to add activity type and site information.
vle_events = student_vle.merge(
    vle,
    on=site_key,
    how="left",
    validate="many_to_one"
)

# Aggregate VLE engagement at the student-course-presentation level.
vle_features = (
    vle_events
    .groupby(student_key, as_index=False)
    .agg(
        vle_rows=("id_site", "count"),
        total_clicks=("sum_click", "sum"),
        active_days=("date", "nunique"),
        first_vle_day=("date", "min"),
        last_vle_day=("date", "max"),
        unique_sites=("id_site", "nunique"),
    )
)

# Calculate average clicks per active day.
vle_features["clicks_per_active_day"] = (
    vle_features["total_clicks"] / vle_features["active_days"]
)

# Preview the generated VLE engagement features.
vle_features.head()

,id_student,code_module,code_presentation,vle_rows,total_clicks,active_days,first_vle_day,last_vle_day,unique_sites,clicks_per_active_day
0,6516,AAA,2014J,662,2791,159,-23,269,84,17.553459
1,8462,DDD,2013J,300,646,56,-6,118,125,11.535714
2,8462,DDD,2014J,4,10,1,10,10,3,10.000000
3,11391,AAA,2013J,196,934,40,-5,253,55,23.350000
4,23629,BBB,2013B,59,161,16,-6,87,11,10.062500


<div style="background:#fff7ed; border:1px solid #fdba74; border-left:5px solid #ea580c; padding:16px 20px; border-radius:12px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<p style="font-size:16px; margin-top:0; color:#c2410c;">
<b>🔎 Observation</b>
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
The VLE features summarize each student's engagement behavior within a module presentation.
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
These features capture important behavioral signals such as total clicks, number of active days, first and last interaction days, number of unique learning resources visited, and average clicks per active day.
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155; margin-bottom:0;">
These engagement indicators are highly relevant for early-warning prediction because reduced or inconsistent VLE activity may indicate risk of failure or withdrawal.
</p>

</div>

<h2 style="color:#EA580C;">🧭 Build VLE Activity-Type Features</h2>

This section breaks down VLE clicks by activity type, such as homepage, forum, quiz, resource, subpage, and URL.

In [21]:
# Create activity-type features by summing clicks for each VLE activity category.
vle_activity_features = (
    vle_events
    .pivot_table(
        index=student_key,
        columns="activity_type",
        values="sum_click",
        aggfunc="sum",
        fill_value=0
    )
)

# Rename activity columns to clear feature names.
vle_activity_features.columns = [
    f"clicks_{column}"
    for column in vle_activity_features.columns
]

# Reset index to keep student identifiers as regular columns.
vle_activity_features = vle_activity_features.reset_index()

# Preview VLE activity-type features.
vle_activity_features.head()

,id_student,code_module,code_presentation


<div style="background:#fff7ed; border:1px solid #fdba74; border-left:5px solid #ea580c; padding:16px 20px; border-radius:12px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<p style="font-size:16px; margin-top:0; color:#c2410c;">
<b>🔎 Observation</b>
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
The VLE activity-type features show how each student's clicks are distributed across different learning resources.
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155; margin-bottom:0;">
This adds more detailed engagement signals than total clicks alone, because students may interact differently with forums, quizzes, resources, homepages, and other learning materials.
</p>

</div>

<h2 style="color:#EA580C;">🧱 Create the Flat Analysis Table</h2>

This section combines student information, registration data, course metadata, assessment features, and VLE engagement features into one student-level analysis table.

In [22]:
# Start from student_info because it already contains one row per student-course-presentation attempt.
# It also includes the target column: final_result.
flat = (
    student_info
    .merge(student_registration, on=student_key, how="left", validate="one_to_one")
    .merge(courses, on=course_key, how="left", validate="many_to_one")
    .merge(assessment_features, on=student_key, how="left", validate="one_to_one")
    .merge(assessment_type_features, on=student_key, how="left", validate="one_to_one")
    .merge(vle_features, on=student_key, how="left", validate="one_to_one")
    .merge(vle_activity_features, on=student_key, how="left", validate="one_to_one")
)

# Identify generated numeric feature columns that should be filled with zero when missing.
feature_columns_to_fill = [
    column for column in flat.columns
    if column.startswith(("assessment_", "vle_", "clicks_"))
    or column in [
        "total_clicks",
        "active_days",
        "unique_sites",
        "late_submissions",
        "banked_assessments"
    ]
]

# Fill missing feature values with zero.
# Missing values here usually mean the student had no recorded assessment or VLE activity.
flat[feature_columns_to_fill] = flat[feature_columns_to_fill].fillna(0)

# Display the final table shape and preview the first rows.
flat.shape, flat.head()

((32593, 37),
   code_module code_presentation  id_student gender                region  \
 0         AAA             2013J       11391      M   East Anglian Region   
 1         AAA             2013J       28400      F              Scotland   
 2         AAA             2013J       30268      F  North Western Region   
 3         AAA             2013J       31604      F     South East Region   
 4         AAA             2013J       32885      F  West Midlands Region   
 
        highest_education imd_band age_band  num_of_prev_attempts  \
 0       HE Qualification  90-100%     55<=                     0   
 1       HE Qualification   20-30%    35-55                     0   
 2  A Level or Equivalent   30-40%    35-55                     0   
 3  A Level or Equivalent   50-60%    35-55                     0   
 4     Lower Than A Level   50-60%     0-35                     0   
 
    studied_credits  ... assessment_mean_cma assessment_mean_exam  \
 0              240  ...             

<div style="background:#fff7ed; border:1px solid #fdba74; border-left:5px solid #ea580c; padding:16px 20px; border-radius:12px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<p style="font-size:16px; margin-top:0; color:#c2410c;">
<b>🔎 Observation</b>
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
The flat analysis table contains one row per <b>student-course-presentation attempt</b> and combines demographic, registration, course, assessment, and VLE engagement features.
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155; margin-bottom:0;">
The resulting table has 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">32,593</code> rows and 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">57</code> columns, making it suitable as a structured machine-learning dataset for later early-warning prediction.
</p>

</div>

In [23]:
# Inspect the final flat table structure, column types, non-null counts, and memory usage.
flat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32593 entries, 0 to 32592
Data columns (total 37 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   code_module                     32593 non-null  object 
 1   code_presentation               32593 non-null  object 
 2   id_student                      32593 non-null  int64  
 3   gender                          32593 non-null  object 
 4   region                          32593 non-null  object 
 5   highest_education               32593 non-null  object 
 6   imd_band                        31482 non-null  object 
 7   age_band                        32593 non-null  object 
 8   num_of_prev_attempts            32593 non-null  int64  
 9   studied_credits                 32593 non-null  int64  
 10  disability                      32593 non-null  object 
 11  final_result                    32593 non-null  object 
 12  date_registration               

<div style="background:#fff7ed; border:1px solid #fdba74; border-left:5px solid #ea580c; padding:16px 20px; border-radius:12px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<p style="font-size:16px; margin-top:0; color:#c2410c;">
<b>🔎 Observation</b>
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
The final flat table contains <b>57 columns</b> with a mix of categorical and numerical features.
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155;">
Most generated assessment and VLE feature columns are fully populated after filling missing feature values with zero. Remaining missing values appear mainly in original contextual columns such as 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">imd_band</code>, 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">date_registration</code>, 
and 
<code style="background:#ffedd5; color:#9a3412; padding:3px 7px; border-radius:6px;">date_unregistration</code>.
</p>

<p style="font-size:15.5px; line-height:1.8; color:#334155; margin-bottom:0;">
These remaining missing values should be handled later during preprocessing before model training.
</p>

</div>

<div style="background:#f8fafc; border:1px solid #dbeafe; border-left:7px solid #2563eb; padding:26px 30px; border-radius:16px; font-family:Segoe UI, Arial, sans-serif; color:#1e293b;">

<h2 style="margin-top:0; color:#1d4ed8; font-size:30px;">
✅ Exploration Summary
</h2>

<p style="font-size:17px; line-height:1.8; color:#334155;">
This notebook explored the raw <b>OULAD tables</b>, checked file structure, missing values, categorical distributions, relational keys, and built an initial flat analysis table for inspection.
</p>

<hr style="border:0; border-top:1px solid #bfdbfe; margin:24px 0;">

<h2 style="color:#0f172a; font-size:26px;">
📦 Next Step
</h2>

<div style="background:#ffffff; border:1px solid #bfdbfe; border-radius:14px; padding:18px 22px; margin:14px 0 22px 0; box-shadow:0 2px 8px rgba(37,99,235,0.08);">

<p style="font-size:16px; line-height:1.8; color:#334155; margin:0;">
The final dataset export will be handled in the next notebook: 
<code style="background:#e0f2fe; color:#075985; padding:3px 7px; border-radius:6px;">02_merge_oulad_data.ipynb</code>.
</p>

</div>

</div>